From 41202931f391c91420622c0301654647bfead311 Mon Sep 17 00:00:00 2001
From: Wen Gong <quic_wgong@quicinc.com>
Date: Mon, 5 Dec 2022 04:35:27 -0500
Subject: [PATCH 043/112] wifi: mac80211: fix recovery fail while firmware
 crash when doing channel switch

When firmware crashed while channel switch running, recovery started in
ath12k, then ieee80211_sta_connection_lost() will be called by function
ieee80211_restart_work() in mac80211, then many wmi command timeout
because firmware is crashed, each WMI command cost 3 seconds, then the
total time will be large and lead recovery fail. Change to set value
ATH12K_FLAG_CRASH_FLUSH early and then ath12k_wmi_cmd_send() will not
wait 3 seconds and recovery will be started quickly and success.

Tested-on: WCN7850 hw2.0 PCI WLAN.HMT.1.0-02582-QCAHMTSWPL_V1.0_V2.0_SILICONZ-1

Signed-off-by: Wen Gong <quic_wgong@quicinc.com>
Signed-off-by: Amutha Ravikumar <quic_aravikum@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 2063a30e7c6e..0b06cc6fc30f 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -743,6 +743,9 @@ static void ath12k_core_pre_reconfigure_recovery(struct ath12k_base *ab)
 	ab->stats.fw_crash_counter++;
 	spin_unlock_bh(&ab->base_lock);
 
+	if (ab->is_reset)
+		set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
+
 	for (i = 0; i < ab->num_radios; i++) {
 		pdev = &ab->pdevs[i];
 		ar = pdev->ar;
-- 
2.31.1

